internal/ls/folding.go

```diff
@@ -0,0 +1,541 @@
+package ls
+
+import (
+	"cmp"
+	"context"
+	"regexp"
+	"slices"
+	"strings"
+	"unicode"
+
+	"github.com/microsoft/typescript-go/internal/ast"
+	"github.com/microsoft/typescript-go/internal/astnav"
+	"github.com/microsoft/typescript-go/internal/lsp/lsproto"
+	"github.com/microsoft/typescript-go/internal/printer"
+	"github.com/microsoft/typescript-go/internal/scanner"
+)
+
+func (l *LanguageService) ProvideFoldingRange(ctx context.Context, documentURI lsproto.DocumentUri) []*lsproto.FoldingRange {
+	_, sourceFile := l.getProgramAndFile(documentURI)
+	res := l.addNodeOutliningSpans(sourceFile)
+	res = append(res, l.addRegionOutliningSpans(sourceFile)...)
+	slices.SortFunc(res, func(a, b *lsproto.FoldingRange) int {
+		if a.StartLine != b.StartLine {
+			return cmp.Compare(a.StartLine, b.StartLine)
+		}
+		if a.StartCharacter != nil && b.StartCharacter != nil {
+			return cmp.Compare(*a.StartCharacter, *b.StartCharacter)
+		}
+		if a.EndLine != b.EndLine {
+			return cmp.Compare(a.EndLine, b.EndLine)
+		}
+		if a.EndCharacter != nil && b.EndCharacter != nil {
+			return cmp.Compare(*a.EndCharacter, *b.EndCharacter)
+		}
+		return 0
+	})
+	return res
+}
+
+func (l *LanguageService) addNodeOutliningSpans(sourceFile *ast.SourceFile) []*lsproto.FoldingRange {
+	depthRemaining := 40
+	current := 0
+
+	statements := sourceFile.Statements
+	n := len(statements.Nodes)
+	var foldingRange []*lsproto.FoldingRange
```

Comment:
I actually think starting out with a nil slice is probably not ideal. Folding ranges are never cached on our side, so maybe it's best to just start out with some reasonable initial capacity (I dunno, I'm just going to throw out a nice round number like 40?).

This isn't a blocker for this PR.